eventual-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/eventual/cassandra/SnapshotCassandra.scala

```diff
@@ -0,0 +1,182 @@
+package com.evolutiongaming.kafka.journal.eventual.cassandra
+
+import cats.effect.kernel.{Async, Resource, Temporal}
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import cats.{Monad, MonadThrow, Parallel}
+import com.evolutiongaming.catshelper.LogOf
+import com.evolutiongaming.kafka.journal._
+import com.evolutiongaming.kafka.journal.eventual.EventualPayloadAndType
+import com.evolutiongaming.kafka.journal.eventual.cassandra.EventualCassandraConfig.ConsistencyConfig
+import com.evolutiongaming.scassandra.CassandraClusterOf
+
+import java.time.Instant
+
+object SnapshotCassandra {
+
+  // TODO: make it configurable
+  val BufferSize = 10
+
+  def of[F[_]: Async: Parallel: LogOf](
+    config: EventualCassandraConfig,
+    origin: Option[Origin],
+    cassandraClusterOf: CassandraClusterOf[F]
+  ): Resource[F, SnapshotStoreFlat[F]] = {
+
+    def store(implicit cassandraCluster: CassandraCluster[F], cassandraSession: CassandraSession[F]) = {
+      of(config.schema, origin, config.consistencyConfig)
+    }
+
+    for {
+      cassandraCluster <- CassandraCluster.of[F](config.client, cassandraClusterOf, config.retries)
+      cassandraSession <- cassandraCluster.session
+      store <- store(cassandraCluster, cassandraSession).toResource
+    } yield store
+
+  }
+
+  def of[F[_]: Temporal: Parallel: CassandraCluster: CassandraSession: LogOf](
+    schemaConfig: SchemaConfig,
+    origin: Option[Origin],
+    consistencyConfig: ConsistencyConfig
+  ): F[SnapshotStoreFlat[F]] =
+    for {
+      schema <- SetupSchema[F](schemaConfig, origin, consistencyConfig)
+      statements <- Statements.of[F](schema, consistencyConfig)
+    } yield SnapshotCassandra(statements)
+
+  private sealed abstract class Main
+
+  def apply[F[_]: MonadThrow](statements: Statements[F]): SnapshotStoreFlat[F] = {
+    new Main with SnapshotStoreFlat[F] {
+
+      // we do not use segments for now
+      val segmentNr = SegmentNr.min
+
+      def save(key: Key, snapshot: SnapshotRecord[EventualPayloadAndType]): F[Unit] = {
+        statements.selectMetadata(key, segmentNr).flatMap {
```

Comment:
on the hot path given that buffer is full, do you need to query ALL snapshots?

Reply:
I need this information to find the oldest snapshot to be replaced. There are might be the ways to optimize it, though. :thinking: